core/scheduler/resolver/lineage_resolver.go

```diff
@@ -0,0 +1,324 @@
+	GetJobRunsByIdentifiers(ctx context.Context, jobRuns []scheduler.JobRunIdentifier) ([]*scheduler.JobRunSummary, error)
+}
+
+type ProjectGetter interface {
+	Get(context.Context, tenant.ProjectName) (*tenant.Project, error)
+}
+
+type LineageResolver struct {
+	upstreamRepo  JobUpstreamRepository
+	jobRepo       JobRepository
+	jobRunService JobRunService
+	projectGetter ProjectGetter
+}
+
+func NewLineageResolver(
+	upstreamRepo JobUpstreamRepository,
+	jobRepo JobRepository,
+	jobRunService JobRunService,
+	projectGetter ProjectGetter,
+) *LineageResolver {
+	return &LineageResolver{
+		upstreamRepo:  upstreamRepo,
+		jobRepo:       jobRepo,
+		jobRunService: jobRunService,
+		projectGetter: projectGetter,
+	}
+}
+
+func (r *LineageResolver) BuildLineage(ctx context.Context, jobSchedules []*scheduler.JobSchedule) ([]*scheduler.JobLineageSummary, error) {
+	lineage, err := r.buildLineageStructure(ctx, jobSchedules)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := r.enrichWithJobDetails(ctx, lineage); err != nil {
+		return nil, err
+	}
+
+	if err := r.enrichWithJobRunDetails(ctx, lineage); err != nil {
+		return nil, err
+	}
+
+	return lineage, nil
+}
+
+func (r *LineageResolver) buildLineageStructure(ctx context.Context, jobSchedules []*scheduler.JobSchedule) ([]*scheduler.JobLineageSummary, error) {
+	queue := []*scheduler.JobLineageSummary{}
+	for _, jobSchedule := range jobSchedules {
+		queue = append(queue, &scheduler.JobLineageSummary{
+			JobName: jobSchedule.JobName,
+			JobRuns: map[string]*scheduler.JobRunSummary{
+				jobSchedule.ScheduledAt.Format(time.RFC3339): {
+					ScheduledAt: jobSchedule.ScheduledAt,
+				},
+			},
+			Upstreams: []*scheduler.JobLineageSummary{},
+		})
+	}
+
+	upstreamsByJob, err := r.upstreamRepo.GetAllResolvedUpstreams(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	jobWithLineages := queue
+	jobWithLineageByNameMap := map[scheduler.JobName]*scheduler.JobLineageSummary{}
+
+	for len(queue) > 0 {
+		jobWithLineage := queue[0]
+		queue = queue[1:]
+
+		for _, upstreamJobName := range upstreamsByJob[jobWithLineage.JobName] {
+			if _, ok := jobWithLineageByNameMap[upstreamJobName]; !ok {
+				upstreamJobWithLineage := &scheduler.JobLineageSummary{
+					JobName:   upstreamJobName,
+					Upstreams: []*scheduler.JobLineageSummary{},
+				}
+				jobWithLineageByNameMap[upstreamJobName] = upstreamJobWithLineage
+			}
+			jobWithLineage.Upstreams = append(jobWithLineage.Upstreams, jobWithLineageByNameMap[upstreamJobName])
+		}
+
+		queue = append(queue, jobWithLineage.Upstreams...)
+	}
+
+	return jobWithLineages, nil
+}
+
+func (r *LineageResolver) enrichWithJobDetails(ctx context.Context, lineage []*scheduler.JobLineageSummary) error {
+	uniqueJobs := collectUniqueJobs(lineage)
+
+	jobNames := make([]scheduler.JobName, 0, len(uniqueJobs))
+	for jobName := range uniqueJobs {
+		jobNames = append(jobNames, jobName)
+	}
+
+	jobsByName, err := r.jobRepo.FindByNames(ctx, jobNames)
```

Comment:
this can be called once, instead of recalling it again in 2 function `enrichWithJobDetails` and `getJobsAndProjects`